File: proto-java/src/main/java/com/spaceape/protobuf/CopyContext.java
package com.spaceape.protobuf;

import java.util.HashMap;
import java.util.Map;

public class CopyContext {
    private Map<GeneratedMessage, GeneratedMessage> copiedMessages = new HashMap<>();

    public GeneratedMessage copyFrom(GeneratedMessage src, ConstructorFactory factory) {
        if (copiedMessages.containsKey(src)) return copiedMessages.get(src);
        GeneratedMessage copiedMsg = factory.newInstance(src.getClass().getName());
        copiedMessages.put(src, copiedMsg);
        copiedMsg.deepCopy(src, this);
        return copiedMsg;
    }
}

File: proto-csharp/ProtoMessage/ProtoMessage/ApeMessage/CopyContext.cs
using System;
using System.Collections.Generic;

namespace com.spaceape.protobuf
{
  public class CopyContext
  {
    private Dictionary<GeneratedMessage, GeneratedMessage> copiedMessages = new Dictionary<GeneratedMessage, GeneratedMessage>();

    public GeneratedMessage copyFrom(GeneratedMessage src, ConstructorFactory factory) {
        if (copiedMessages.ContainsKey(src)) return copiedMessages[src];
        GeneratedMessage copiedMsg = factory.newInstance(src.GetType().FullName);
        copiedMessages.Add(src, copiedMsg);
        copiedMsg.deepCopy(src, this);
        return copiedMsg;
    }
  }
}

